lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp

```diff
@@ -0,0 +1,893 @@
+    return static_cast<CombNode>(pointer);
+  }
+  static CombNode getTombstoneKey() {
+    auto pointer = llvm::DenseMapInfo<void *>::getTombstoneKey();
+    return static_cast<CombNode>(pointer);
+  }
+
+  static unsigned getHashValue(const CombNodeImpl *node) {
+    return node->hash_value();
+  }
+  static bool isEqual(const CombNodeImpl *lhs, const CombNodeImpl *rhs) {
+    auto empty = getEmptyKey();
+    auto tombstone = getTombstoneKey();
+    if (lhs == empty || lhs == tombstone || rhs == empty || rhs == tombstone)
+      return lhs == rhs;
+    return *lhs == *rhs;
+  }
+
+  // Allow lookup via find_as(), without constructing a temporary CombNodeImpl.
+  static unsigned getHashValue(const CombNodeKeyType key) {
+    return llvm::hash_combine(key.first.getOpaqueValue(), key.second);
+  }
+  static bool isEqual(const CombNodeKeyType lhs, const CombNodeImpl *rhs) {
+    auto empty = getEmptyKey();
+    auto tombstone = getTombstoneKey();
+    if (rhs == empty || rhs == tombstone)
+      return false;
+    return lhs.first == rhs->getValOrOp() && lhs.second == rhs->getFieldID();
+  }
+};
+} // namespace llvm
+
+//===----------------------------------------------------------------------===//
+// SSA graph based implementation
+//===----------------------------------------------------------------------===//
+
+using CombPathsType = SmallVector<SmallVector<size_t, 2>>;
+using CombPathsMapType = DenseMap<Operation *, CombPathsType>;
+
+namespace {
+/// The graph context containing pointers of the combinational paths map and the
+/// instance graph.
+struct GraphContext {
+  CombPathsMapType *map;
+  InstanceGraph *graph;
+
+  explicit GraphContext(CombPathsMapType *map, InstanceGraph *graph)
+      : map(map), graph(graph) {}
+};
+} // namespace
+
+namespace {
+/// The node class of combinational graph.
+struct Node {
+  Value value;
+  GraphContext *context;
+
+  explicit Node(Value value = nullptr, GraphContext *context = nullptr)
+      : value(value), context(context) {}
+
+  bool operator==(const Node &rhs) const { return value == rhs.value; }
+  bool operator!=(const Node &rhs) const { return !(*this == rhs); }
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// ChildIterator class
+//===----------------------------------------------------------------------===//
+
+using use_iterator =
+    llvm::filter_iterator<Value::use_iterator, bool (*)(OpOperand &)>;
+using child_iterator =
+    llvm::mapped_iterator<use_iterator, Value (*)(OpOperand &)>;
+
+namespace {
+/// This is a child iterator class for FIRRTL values. The base iterator is the
+/// `use_iterator` of a value. On top of the `use_iterator`, uses whose owner
+/// has no results and uses which are 'dest`s of connect ops are filtered out
+/// through the `isChild` method. Then, all the filtered uses are mapped to its
+/// corresponding child value through the `getChild` method.
+class ChildIterator : public child_iterator {
```

Comment:
IMO, the use of filter_iterator/mapped_iterator just makes this more complicated.  I'd recommend flattening their implementations into ChildIterator.    C++ isn't great at composing functional algorithms in this way.

Reply:
Thanks for the suggestion! After dropping `filter_iterator` and `mapped_iterator`, the code is much cleaner and easy to follow now :smile: